xds/src/main/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProvider.java

```diff
@@ -17,35 +17,346 @@
+      try {
+        Meshca.MeshCertificateResponse response =
+            stub.withDeadlineAfter(rpcTimeoutMillis, TimeUnit.MILLISECONDS)
+                .createCertificate(request);
+        return getX509CertificatesFromResponse(response);
+      } catch (Throwable t) {
+        if (!retriable(t)) {
+          generateErrorIfCurrentCertExpired(t);
+          return null;
+        }
+        lastException = t;
+        sleepForNanos(backoffPolicy.nextBackoffNanos());
+      }
+    }
+    generateErrorIfCurrentCertExpired(lastException);
+    return null;
+  }
+
+  private void sleepForNanos(long nanos) {
+    ScheduledFuture<?> future = scheduledExecutorService.schedule(new Runnable() {
+      @Override
+      public void run() {
+        // do nothing
+      }
+    }, nanos, TimeUnit.NANOSECONDS);
+    try {
+      future.get(nanos, TimeUnit.NANOSECONDS);
+    } catch (InterruptedException ie) {
+      logger.log(Level.SEVERE, "Inside sleep", ie);
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException | TimeoutException ex) {
+      logger.log(Level.SEVERE, "Inside sleep", ex);
+    }
+  }
+
+  private static boolean retriable(Throwable t) {
+    if (t instanceof StatusRuntimeException) {
+      return retriable(((StatusRuntimeException) t).getStatus());
+    } else if (t.getCause() instanceof StatusRuntimeException) {
+      return retriable(((StatusRuntimeException) t.getCause()).getStatus());
+    }
+    return false;
+  }
+
+  private static boolean retriable(Status status) {
+    return RETRIABLE_CODES.contains(status.getCode());
+  }
+
+  private void generateErrorIfCurrentCertExpired(Throwable t) {
+    X509Certificate currentCert = getWatcher().getLastIdentityCert();
+    if (currentCert != null) {
+      long delaySeconds = computeDelaySecondsToCertExpiry(currentCert);
+      if (delaySeconds > INITIAL_DELAY_SECONDS) {
+        return;
+      }
+      getWatcher().clearValues();
+    }
+    getWatcher().onError(Status.fromThrowable(t));
+  }
+
+  private KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(alg);
+    keyPairGenerator.initialize(keySize);
+    return keyPairGenerator.generateKeyPair();
+  }
+
+  private String generateCsr(KeyPair pair) throws IOException, OperatorCreationException {
+    PKCS10CertificationRequestBuilder p10Builder =
+        new JcaPKCS10CertificationRequestBuilder(
+            new X500Principal("CN=EXAMPLE.COM"), pair.getPublic());
+    JcaContentSignerBuilder csBuilder = new JcaContentSignerBuilder(signatureAlg);
+    ContentSigner signer = csBuilder.build(pair.getPrivate());
+    PKCS10CertificationRequest csr = p10Builder.build(signer);
+    PemObject pemObject = new PemObject("NEW CERTIFICATE REQUEST", csr.getEncoded());
+    StringWriter str = new StringWriter();
+    JcaPEMWriter pemWriter = new JcaPEMWriter(str);
+    pemWriter.writeObject(pemObject);
+    pemWriter.close();
+    str.close();
```

Comment:
Seems we can use java7's try-with-resource now.

Reply:
Okay, BTW a single try-with-resources like this doesn't work:
```
    try (StringWriter str = new StringWriter(); JcaPEMWriter pemWriter = new JcaPEMWriter(str)) {
      pemWriter.writeObject(pemObject);
      return str.toString();
    }
```
but this one works:
```
    try (StringWriter str = new StringWriter()) {
      try (JcaPEMWriter pemWriter = new JcaPEMWriter(str)) {
        pemWriter.writeObject(pemObject);
      }
      return str.toString();
    }
```